File: 416-partition-equal-subset-sum/416-partition-equal-subset-sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = accumulate(nums.begin(), nums.end(), 0);
        if((sum & 1) == 1) return false;
        return solve(nums, nums.size(), sum/2);
    }
private:
    bool solve(vector<int> nums, int n, int k){
        vector<vector<bool>> dp(n+1, vector<bool>(k+1, false));
        
        for(int i = 0; i <= n; ++i)
            dp[i][0] = true;
        
        for(int i = 1; i <= n; ++i) {
            for(int j = 1; j <= k; ++j) {
                if(nums[i-1] <= j)
                    dp[i][j] = dp[i-1][j] || dp[i-1][j - nums[i-1]];
                else
                    dp[i][j] = dp[i-1][j];
            }
        }
        return dp[n][k];
    }
};

File: 416-partition-equal-subset-sum/416-partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        
        def solve(arr, n, k):
            dp = [[False]*(k+1) for _ in range(n+1)]
            
            for i in range(n+1):
                dp[i][0] = True
                
            for i in range(1, n+1):
                for j in range(1, k+1):
                    if arr[i-1] > j:
                        dp[i][j] = dp[i-1][j]
                    else:
                        dp[i][j] = dp[i-1][j] or dp[i-1][j-arr[i-1]]
            return dp[n][k] 
        
        add = sum(nums)
        if (add & 1) == 1: return False
        return solve(nums, len(nums), add//2)